File: python/eou/special_trie.py
from __future__ import annotations

from dataclasses import dataclass, field

@dataclass
class SpecialPiece:
    text: str
    id: int
    special: bool

class SpecialNode:
    __slots__ = ("children", "terminal", "id", "content")

    def __init__(self) -> None:
        self.children: dict[int, "SpecialNode"] = {}
        self.terminal: bool = False
        self.id: int = 0
        self.content: str = ""

    def insert(self, s: str, id: int) -> None:
        cur = self
        for b in s.encode("utf-8"):
            nxt = cur.children.get(b)
            if nxt is None:
                nxt = SpecialNode()
                cur.children[b] = nxt
            cur = nxt
        cur.terminal = True
        cur.id = id
        cur.content = s

    def split(self, text: str) -> list[SpecialPiece]:
        data = text.encode("utf-8")
        out: list[SpecialPiece] = []
        plain = bytearray()
        i = 0
        n = len(data)

        def flush() -> None:
            if plain:
                out.append(
                    SpecialPiece(
                        text=plain.decode("utf-8", errors="replace"),
                        id=-1,
                        special=False,
                    )
                )
                plain.clear()

        while i < n:
            mid, mlen, mtext = self._match_at(data, i)
            if mlen > 0:
                flush()
                out.append(SpecialPiece(text=mtext, id=mid, special=True))
                i += mlen
            else:
                plain.append(data[i])
                i += 1
        flush()
        return out

    def _match_at(self, data: bytes, start: int) -> tuple[int, int, str]:
        cur = self
        match_id = -1
        match_end = 0
        match_text = ""
        i = start
        n = len(data)
        while i < n:
            nxt = cur.children.get(data[i])
            if nxt is None:
                break
            cur = nxt
            if cur.terminal:
                match_id = cur.id
                match_end = i + 1
                match_text = cur.content
            i += 1
        if match_end == 0:
            return (-1, 0, "")
        return (match_id, match_end - start, match_text)

File: rust/src/eou/special_trie.rs
#![allow(dead_code)]

use std::collections::HashMap;

#[derive(Default)]
pub struct SpecialNode {
    children: HashMap<u8, SpecialNode>,
    terminal: bool,
    id: i64,
    content: String,
}

#[derive(Debug, Clone)]
pub struct SpecialPiece {
    pub text: String,
    pub id: i64,
    pub special: bool,
}

impl SpecialNode {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert(&mut self, s: &str, id: i64) {
        let mut cur = self;
        for b in s.as_bytes() {
            cur = cur.children.entry(*b).or_default();
        }
        cur.terminal = true;
        cur.id = id;
        cur.content = s.to_string();
    }

    pub fn split(&self, text: &str) -> Vec<SpecialPiece> {
        let bytes = text.as_bytes();
        let mut out: Vec<SpecialPiece> = Vec::new();
        let mut plain: Vec<u8> = Vec::new();
        let mut i = 0usize;
        let flush = |plain: &mut Vec<u8>, out: &mut Vec<SpecialPiece>| {
            if !plain.is_empty() {
                let s = String::from_utf8_lossy(plain).into_owned();
                out.push(SpecialPiece {
                    text: s,
                    id: -1,
                    special: false,
                });
                plain.clear();
            }
        };
        while i < bytes.len() {
            let (id, mlen, mtext) = self.match_at(bytes, i);
            if mlen > 0 {
                flush(&mut plain, &mut out);
                out.push(SpecialPiece {
                    text: mtext,
                    id,
                    special: true,
                });
                i += mlen;
            } else {
                plain.push(bytes[i]);
                i += 1;
            }
        }
        flush(&mut plain, &mut out);
        out
    }

    fn match_at(&self, bytes: &[u8], start: usize) -> (i64, usize, String) {
        let mut cur = self;
        let mut match_id: i64 = -1;
        let mut match_end: usize = 0;
        let mut match_text = String::new();
        let mut i = start;
        while i < bytes.len() {
            match cur.children.get(&bytes[i]) {
                Some(next) => {
                    cur = next;
                    if cur.terminal {
                        match_id = cur.id;
                        match_end = i + 1;
                        match_text = cur.content.clone();
                    }
                    i += 1;
                }
                None => break,
            }
        }
        if match_end == 0 {
            return (-1, 0, String::new());
        }
        (match_id, match_end - start, match_text)
    }
}
